heat/core/devices.py

```diff
@@ -0,0 +1,92 @@
+import torch
+
+from .communication import MPI_WORLD
+
+
+def cpu_index():
+    """
+    Retrieves the index of the CPU to be used for this MPI process.
+
+    Returns
+    -------
+    cpu_index : int
+        The CPU index
+    """
+    return 'cpu'
+
+
+__default_device = 'cpu'
+__device_mapping = {
+    'cpu': cpu_index
+}
+
+# add gpu support if available
+if torch.cuda.device_count() > 0:
+    def gpu_index():
+        """
+        Retrieves the index of the GPU to be used for this MPI process.
+
+        Returns
+        -------
+        gpu_index : int
+            The GPU index
+        """
+        return 'cuda:{}'.format(MPI_WORLD.rank % torch.cuda.device_count())
+
+    __device_mapping['gpu'] = gpu_index
```

Comment:
I do not see why we have to store a function object here, since the return value of the latter is not subject to change?

Reply:
Meant for extensibility. Makes it easier to add new backends without having to touch on big if-else or switch-case function that knows all the backends.

For the GPU backend I can for example dynamically check whether it is available or not.

Should stay as is in my opinion.